nx/lib/nx/defn/grad.ex

```diff
@@ -5,12 +5,19 @@ defmodule Nx.Defn.Grad do
   alias Nx.Tensor, as: T
 
   def transform(to_grad, fun, transform) do
-    {to_grad, ids} =
-      Composite.traverse(to_grad, %{}, fn to_grad, ids ->
+    broadcasted_nodes =
+      [to_grad]
+      |> Composite.flatten_list()
+      |> Nx.broadcast_vectors()
+
+    {to_grad, {ids, []}} =
+      Composite.traverse(to_grad, {%{}, broadcasted_nodes}, fn _to_grad,
+                                                               {ids, [broadcasted_node | nodes]} ->
         to_grad =
-          Expr.metadata(to_grad, %{__MODULE__ => :to_grad})
+          Expr.metadata(broadcasted_node, %{__MODULE__ => :to_grad})
 
-        {to_grad, Map.put(ids, to_grad.data.id, :stop)}
+        ids = Map.put(ids, to_grad.data.id, :stop)
+        {to_grad, {ids, nodes}}
       end)
```

Comment:
This isn't actually the proper fix. This solves the issue only when there is no closure for a vectorized tensor.
fn y -> Nx.add(x, y) end fails if y is unvectorized but x is vectorized